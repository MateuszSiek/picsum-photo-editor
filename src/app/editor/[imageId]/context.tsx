import { RefObject, createContext, useRef } from 'react';

export interface CanvasContextProps {
  canvasRef?: RefObject<HTMLCanvasElement>;
}

export const CanvasContext = createContext<CanvasContextProps>({});

export const CanvasContextProvider = ({ children }: any) => {
  const canvasRef = useRef<HTMLCanvasElement>(null);

  return (
    <CanvasContext.Provider value={{ canvasRef }}>
      {children}
    </CanvasContext.Provider>
  );
};
