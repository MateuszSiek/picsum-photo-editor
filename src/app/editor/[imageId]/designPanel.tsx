import {
  LabeledInput,
  LabeledSlider,
  LabeledSwitch,
} from '@/components/labeledInput';
import { Button } from '@/components/ui/button';
import { loadPicsumImage } from '@/lib/picsumApi';
import { PicsumImage } from '@/lib/types';
import { DownloadIcon } from '@radix-ui/react-icons';
import { useContext, useEffect, useState } from 'react';
import { CanvasContext } from './context';
import {
  useQueryBlur,
  useQueryGrayScale,
  useQueryImageSize,
} from '@/lib/hooks';

function ImageSize() {
  const [{ width, height }, setSize] = useQueryImageSize();

  return (
    <div className='flex gap-2'>
      <LabeledInput
        className='flex-1'
        onChange={(e) => setSize({ width: Number(e.target.value), height })}
        type='number'
        min={0}
        value={width}
        label={'Width'}
      />
      <LabeledInput
        className='flex-1'
        onChange={(e) => setSize({ height: Number(e.target.value), width })}
        type='number'
        min={0}
        value={height}
        label={'Height'}
      />
    </div>
  );
}

function ImageBlur() {
  const [blur, setBlur] = useQueryBlur();
  return (
    <LabeledSlider
      max={10}
      min={0}
      step={1}
      onValueChange={(v: number[]) => setBlur(v[0])}
      value={[blur]}
      label={'Blur'}
    />
  );
}

function ImageGrayscale() {
  const [grayscale, setGrayscale] = useQueryGrayScale();

  return (
    <LabeledSwitch
      checked={grayscale}
      onCheckedChange={(v) => setGrayscale(v)}
      label={'Grayscale'}
    />
  );
}

function DonwloadButton() {
  const { canvasRef, canvasReady } = useContext(CanvasContext);

  const downloadImage = () => {
    if (canvasRef) {
      const link = document.createElement('a');
      link.download = 'image.png';
      link.href = canvasRef.current?.toDataURL('image/png')!;
      link.click();
      link.remove();
    }
  };

  return (
    <Button disabled={!canvasReady} onClick={downloadImage} className='w-full'>
      <DownloadIcon className='mr-2 h-5 w-5' /> Download
    </Button>
  );
}

export function DesignPanel({ image }: { image?: PicsumImage }) {
  const [{ width, height }, setSize] = useQueryImageSize(false);

  useEffect(() => {
    if (!image) return;
    if (!width || !height)
      setSize(
        { width: image.width, height: image.height },
        { history: 'replace' }
      );
  }, [image]);

  return (
    <div className='flex flex-col'>
      <ImageSize />
      <div className='pt-4'>
        <ImageBlur />
      </div>
      <div className='pt-8'>
        <ImageGrayscale />
      </div>
      <div className='pt-6'>
        <DonwloadButton />
      </div>
    </div>
  );
}
