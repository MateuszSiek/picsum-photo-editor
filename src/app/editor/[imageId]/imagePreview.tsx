import {
  useQueryBlur,
  useQueryGrayScale,
  useViewScale,
  useQueryImageSize,
} from '@/lib/hooks';
import { PicsumImage } from '@/lib/types';
import { useContext, useEffect, useRef, useState } from 'react';
import { CanvasContext } from './context';
import { LoadingSpinner } from '@/components/ui/spinner';

export function calculateDrawParameters(
  canvasWidth: number,
  canvasHeight: number,
  imageWidth: number,
  imageHeight: number
) {
  const canvasAspect = canvasWidth / canvasHeight;
  const imageAspect = imageWidth / imageHeight;
  let drawWidth, drawHeight, offsetX, offsetY;
  if (canvasAspect > imageAspect) {
    drawWidth = canvasWidth;
    drawHeight = canvasWidth / imageAspect;
    offsetX = 0;
    offsetY = (canvasHeight - drawHeight) / 2;
  } else {
    drawWidth = canvasHeight * imageAspect;
    drawHeight = canvasHeight;
    offsetX = (canvasWidth - drawWidth) / 2;
    offsetY = 0;
  }
  return { drawWidth, drawHeight, offsetX, offsetY };
}

function ImageCanvas({
  image,
  width,
  height,
  style,
}: {
  image?: PicsumImage;
  width: number;
  height: number;
  style: React.CSSProperties;
}) {
  const { canvasRef, setCanvasReady } = useContext(CanvasContext);
  const [canvasImage, setCanvasImage] = useState<HTMLImageElement | null>(null);
  const [grayscale] = useQueryGrayScale();
  const [blur] = useQueryBlur();

  useEffect(() => {
    // https://issues.chromium.org/issues/328755781
    document.addEventListener('visibilitychange', function () {
      if (document.visibilityState === 'visible') {
        const context = canvasRef?.current?.getContext('2d');
        if (context) {
          context.fillStyle = 'transparent';
          context.fillRect(0, 0, 1, 1);
        }
      }
    });
  }, [canvasRef]);

  useEffect(() => {
    if (image) {
      const img = new Image();
      img.src = image.download_url;
      img.onload = () => setCanvasImage(img);
      img.setAttribute('crossOrigin', 'anonymous');
    }
  }, [image]);

  useEffect(() => {
    // Draw the image with filters on the canvas
    const ctx = canvasRef?.current?.getContext('2d');
    if (canvasImage && image && ctx) {
      ctx.clearRect(0, 0, width, height);

      let filter = 'none';
      if (grayscale || blur) {
        filter = `${grayscale ? 'grayscale(100%)' : ''} ${blur ? `blur(${blur}px)` : ''}`;
      }
      ctx.filter = filter;
      const { drawWidth, drawHeight, offsetX, offsetY } =
        calculateDrawParameters(width, height, image.width, image.height);
      ctx.drawImage(canvasImage, offsetX, offsetY, drawWidth, drawHeight);
      setCanvasReady(true);
    }
  }, [
    canvasImage,
    image,
    width,
    height,
    blur,
    grayscale,
    canvasRef,
    setCanvasReady,
  ]);

  return (
    <>
      {!canvasImage && (
        <LoadingSpinner className='absolute left-0 right-0 top-20 m-auto' />
      )}
      <canvas
        className='absolute left-0 right-0 m-auto rounded-md'
        style={style}
        ref={canvasRef}
        width={width}
        height={height}
      />
    </>
  );
}

export default function ImagePreview({ image }: { image?: PicsumImage }) {
  const canvasWrapperRef = useRef<HTMLDivElement>(null);
  const [{ width, height }] = useQueryImageSize();

  const canvasScale = useViewScale(canvasWrapperRef, [width, height]);

  return (
    <div ref={canvasWrapperRef} className='relative'>
      <ImageCanvas
        image={image}
        width={width}
        height={height}
        style={{
          width: `${width * canvasScale}px`,
          height: `${height * canvasScale}px`,
        }}
      />
    </div>
  );
}
