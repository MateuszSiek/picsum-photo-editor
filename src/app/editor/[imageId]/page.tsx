'use client';

import { useRouter } from 'next/navigation';
import { PicsumImage } from '@/lib/types';
import { CanvasContextProvider } from './context';
import { DesignPanel } from './designPanel';
import ImagePreview from './imagePreview';
import { useEffect, useState } from 'react';
import { loadPicsumImage } from '@/lib/picsumApi';

export default function Editor({ params }: { params: { imageId: string } }) {
  const imageId = params.imageId;
  const [image, setImage] = useState<PicsumImage>();
  const router = useRouter();

  useEffect(() => {
    loadPicsumImage(imageId)
      .then(setImage)
      .catch((e) => {
        router.replace('/404');
      });
  }, [imageId]);

  return (
    <div className='grid h-full flex-1 items-stretch gap-6 md:grid-cols-[1fr_250px]'>
      <CanvasContextProvider>
        <ImagePreview image={image} />
        <DesignPanel image={image} />
      </CanvasContextProvider>
    </div>
  );
}
