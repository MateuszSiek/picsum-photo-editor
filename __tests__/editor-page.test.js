import '@testing-library/jest-dom';
import 'jest-canvas-mock';
import { render, screen } from '@testing-library/react';
import { act } from 'react';
import Editor from '../src/app/editor/[imageId]/page';
import { calculateDrawParameters } from '../src/app/editor/[imageId]/imagePreview';
import { useRouter } from 'next/navigation';

const mockImage = [
  {
    id: 1,
    author: 'Mock Author 1',
    width: 1000,
    height: 1000,
    url: 'mock.url/1',
    download_url: 'mock-download.url/1',
  },
];

jest.mock('../src/lib/hooks', () => ({
  useQueryImageSize: () => [{ width: 400, height: 300 }, jest.fn()],
  useQueryBlur: () => [4, jest.fn()],
  useQueryGrayScale: () => [true, jest.fn()],
  useViewScale: jest.fn((x) => 10),
}));

global.ResizeObserver = jest.fn().mockImplementation(() => ({
  observe: jest.fn(),
  unobserve: jest.fn(),
  disconnect: jest.fn(),
}));

jest.mock('../src/lib/picsumApi', () => ({
  loadPicsumImage: () => new Promise((resolve) => resolve(mockImage)),
  getScaledPicsumImageURL: (id, width, height) =>
    `mock-scaled.url/${id}/${width}/${height}`,
}));

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}));

describe('Page', () => {
  it('Render images', async () => {
    const editorPage = await act(async () =>
      render(<Editor params={{ imageId: 1 }} />)
    );

    expect(editorPage).toMatchSnapshot();
  });
});

describe('imagePreview/calculateDrawParameters', () => {
  it('calculates correct parameters when canvas and image have same aspect ratio', () => {
    const result = calculateDrawParameters(800, 600, 400, 300);
    const expected = {
      drawWidth: 800,
      drawHeight: 600,
      offsetX: 0,
      offsetY: 0,
    };
    expect(result).toEqual(expected);
  });

  it('calculates correct parameters when canvas is wider than image', () => {
    const result = calculateDrawParameters(1200, 600, 800, 600);
    const expected = {
      drawWidth: 1200,
      drawHeight: 900,
      offsetX: 0,
      offsetY: -150,
    };
    expect(result).toEqual(expected);
  });

  it('calculates correct parameters when canvas is taller than image', () => {
    const result = calculateDrawParameters(600, 1200, 600, 800);
    const expected = {
      drawWidth: 900,
      drawHeight: 1200,
      offsetX: -150,
      offsetY: 0,
    };
    expect(result).toEqual(expected);
  });

  it('calculates correct parameters when image is wider than canvas', () => {
    const result = calculateDrawParameters(800, 600, 1600, 600);
    const expected = {
      drawWidth: 1600,
      drawHeight: 600,
      offsetX: -400,
      offsetY: 0,
    };
    expect(result).toEqual(expected);
  });

  it('calculates correct parameters when image is taller than canvas', () => {
    const result = calculateDrawParameters(600, 800, 600, 1600);
    const expected = {
      drawWidth: 600,
      drawHeight: 1600,
      offsetX: 0,
      offsetY: -400,
    };
    expect(result).toEqual(expected);
  });

  it('handles edge case where canvas and image sizes are equal', () => {
    const result = calculateDrawParameters(1000, 1000, 1000, 1000);
    const expected = {
      drawWidth: 1000,
      drawHeight: 1000,
      offsetX: 0,
      offsetY: 0,
    };
    expect(result).toEqual(expected);
  });
});
