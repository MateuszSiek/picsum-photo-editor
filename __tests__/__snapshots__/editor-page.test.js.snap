// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Render images 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="grid h-full flex-1 items-stretch gap-6 md:grid-cols-[1fr_250px]"
      >
        <div
          class="relative"
        >
          <svg
            class="animate-spin absolute left-0 right-0 top-20 m-auto"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21 12a9 9 0 1 1-6.219-8.56"
            />
          </svg>
          <canvas
            class="absolute left-0 right-0 m-auto rounded-md"
            height="300"
            style="width: 4000px; height: 3000px;"
            width="400"
          />
        </div>
        <div
          class="flex flex-col"
        >
          <div
            class="flex gap-2"
          >
            <div
              class="grid gap-2 flex-1"
            >
              <label
                class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
              >
                Width
              </label>
              <input
                class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50"
                min="0"
                type="number"
                value="400"
              />
            </div>
            <div
              class="grid gap-2 flex-1"
            >
              <label
                class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
              >
                Height
              </label>
              <input
                class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50"
                min="0"
                type="number"
                value="300"
              />
            </div>
          </div>
          <div
            class="pt-4"
          >
            <div
              class="grid gap-2"
            >
              <div
                class="flex items-center justify-between"
              >
                <label
                  class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
                >
                  Blur
                </label>
                <span
                  class="w-12 rounded-md border border-transparent px-2 py-0.5 text-right text-sm text-muted-foreground hover:border-border"
                >
                  4
                </span>
              </div>
              <span
                aria-disabled="false"
                class="relative flex w-full touch-none select-none items-center [&_[role=slider]]:h-4 [&_[role=slider]]:w-4"
                data-orientation="horizontal"
                dir="ltr"
                style="--radix-slider-thumb-transform: translateX(-50%);"
              >
                <span
                  class="relative h-1.5 w-full grow overflow-hidden rounded-full bg-primary/20"
                  data-orientation="horizontal"
                >
                  <span
                    class="absolute h-full bg-primary"
                    data-orientation="horizontal"
                    style="left: 0%; right: 60%;"
                  />
                </span>
                <span
                  style="transform: var(--radix-slider-thumb-transform); position: absolute; left: calc(40% + 0px);"
                >
                  <span
                    aria-orientation="horizontal"
                    aria-valuemax="10"
                    aria-valuemin="0"
                    aria-valuenow="4"
                    class="block h-4 w-4 rounded-full border border-primary/50 bg-background shadow transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50"
                    data-orientation="horizontal"
                    data-radix-collection-item=""
                    role="slider"
                    style=""
                    tabindex="0"
                  />
                </span>
              </span>
            </div>
          </div>
          <div
            class="pt-8"
          >
            <div
              class="flex items-center justify-between"
            >
              <label
                class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
              >
                Grayscale
              </label>
              <button
                aria-checked="true"
                class="peer inline-flex h-5 w-9 shrink-0 cursor-pointer items-center rounded-full border-2 border-transparent shadow-sm transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 focus-visible:ring-offset-background disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=unchecked]:bg-input"
                data-state="checked"
                role="switch"
                type="button"
                value="on"
              >
                <span
                  class="pointer-events-none block h-4 w-4 rounded-full bg-background shadow-lg ring-0 transition-transform data-[state=checked]:translate-x-4 data-[state=unchecked]:translate-x-0"
                  data-state="checked"
                />
              </button>
            </div>
          </div>
          <div
            class="pt-6"
          >
            <button
              class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2 w-full"
              disabled=""
            >
              <svg
                class="mr-2 h-5 w-5"
                fill="none"
                height="15"
                viewBox="0 0 15 15"
                width="15"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M7.50005 1.04999C7.74858 1.04999 7.95005 1.25146 7.95005 1.49999V8.41359L10.1819 6.18179C10.3576 6.00605 10.6425 6.00605 10.8182 6.18179C10.994 6.35753 10.994 6.64245 10.8182 6.81819L7.81825 9.81819C7.64251 9.99392 7.35759 9.99392 7.18185 9.81819L4.18185 6.81819C4.00611 6.64245 4.00611 6.35753 4.18185 6.18179C4.35759 6.00605 4.64251 6.00605 4.81825 6.18179L7.05005 8.41359V1.49999C7.05005 1.25146 7.25152 1.04999 7.50005 1.04999ZM2.5 10C2.77614 10 3 10.2239 3 10.5V12C3 12.5539 3.44565 13 3.99635 13H11.0012C11.5529 13 12 12.5528 12 12V10.5C12 10.2239 12.2239 10 12.5 10C12.7761 10 13 10.2239 13 10.5V12C13 13.1041 12.1062 14 11.0012 14H3.99635C2.89019 14 2 13.103 2 12V10.5C2 10.2239 2.22386 10 2.5 10Z"
                  fill="currentColor"
                  fill-rule="evenodd"
                />
              </svg>
               Download
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="grid h-full flex-1 items-stretch gap-6 md:grid-cols-[1fr_250px]"
    >
      <div
        class="relative"
      >
        <svg
          class="animate-spin absolute left-0 right-0 top-20 m-auto"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21 12a9 9 0 1 1-6.219-8.56"
          />
        </svg>
        <canvas
          class="absolute left-0 right-0 m-auto rounded-md"
          height="300"
          style="width: 4000px; height: 3000px;"
          width="400"
        />
      </div>
      <div
        class="flex flex-col"
      >
        <div
          class="flex gap-2"
        >
          <div
            class="grid gap-2 flex-1"
          >
            <label
              class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
            >
              Width
            </label>
            <input
              class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50"
              min="0"
              type="number"
              value="400"
            />
          </div>
          <div
            class="grid gap-2 flex-1"
          >
            <label
              class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
            >
              Height
            </label>
            <input
              class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50"
              min="0"
              type="number"
              value="300"
            />
          </div>
        </div>
        <div
          class="pt-4"
        >
          <div
            class="grid gap-2"
          >
            <div
              class="flex items-center justify-between"
            >
              <label
                class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
              >
                Blur
              </label>
              <span
                class="w-12 rounded-md border border-transparent px-2 py-0.5 text-right text-sm text-muted-foreground hover:border-border"
              >
                4
              </span>
            </div>
            <span
              aria-disabled="false"
              class="relative flex w-full touch-none select-none items-center [&_[role=slider]]:h-4 [&_[role=slider]]:w-4"
              data-orientation="horizontal"
              dir="ltr"
              style="--radix-slider-thumb-transform: translateX(-50%);"
            >
              <span
                class="relative h-1.5 w-full grow overflow-hidden rounded-full bg-primary/20"
                data-orientation="horizontal"
              >
                <span
                  class="absolute h-full bg-primary"
                  data-orientation="horizontal"
                  style="left: 0%; right: 60%;"
                />
              </span>
              <span
                style="transform: var(--radix-slider-thumb-transform); position: absolute; left: calc(40% + 0px);"
              >
                <span
                  aria-orientation="horizontal"
                  aria-valuemax="10"
                  aria-valuemin="0"
                  aria-valuenow="4"
                  class="block h-4 w-4 rounded-full border border-primary/50 bg-background shadow transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50"
                  data-orientation="horizontal"
                  data-radix-collection-item=""
                  role="slider"
                  style=""
                  tabindex="0"
                />
              </span>
            </span>
          </div>
        </div>
        <div
          class="pt-8"
        >
          <div
            class="flex items-center justify-between"
          >
            <label
              class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
            >
              Grayscale
            </label>
            <button
              aria-checked="true"
              class="peer inline-flex h-5 w-9 shrink-0 cursor-pointer items-center rounded-full border-2 border-transparent shadow-sm transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 focus-visible:ring-offset-background disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=unchecked]:bg-input"
              data-state="checked"
              role="switch"
              type="button"
              value="on"
            >
              <span
                class="pointer-events-none block h-4 w-4 rounded-full bg-background shadow-lg ring-0 transition-transform data-[state=checked]:translate-x-4 data-[state=unchecked]:translate-x-0"
                data-state="checked"
              />
            </button>
          </div>
        </div>
        <div
          class="pt-6"
        >
          <button
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2 w-full"
            disabled=""
          >
            <svg
              class="mr-2 h-5 w-5"
              fill="none"
              height="15"
              viewBox="0 0 15 15"
              width="15"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M7.50005 1.04999C7.74858 1.04999 7.95005 1.25146 7.95005 1.49999V8.41359L10.1819 6.18179C10.3576 6.00605 10.6425 6.00605 10.8182 6.18179C10.994 6.35753 10.994 6.64245 10.8182 6.81819L7.81825 9.81819C7.64251 9.99392 7.35759 9.99392 7.18185 9.81819L4.18185 6.81819C4.00611 6.64245 4.00611 6.35753 4.18185 6.18179C4.35759 6.00605 4.64251 6.00605 4.81825 6.18179L7.05005 8.41359V1.49999C7.05005 1.25146 7.25152 1.04999 7.50005 1.04999ZM2.5 10C2.77614 10 3 10.2239 3 10.5V12C3 12.5539 3.44565 13 3.99635 13H11.0012C11.5529 13 12 12.5528 12 12V10.5C12 10.2239 12.2239 10 12.5 10C12.7761 10 13 10.2239 13 10.5V12C13 13.1041 12.1062 14 11.0012 14H3.99635C2.89019 14 2 13.103 2 12V10.5C2 10.2239 2.22386 10 2.5 10Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
             Download
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
